src/mgr/QueryParams.cc

```diff
@@ -66,60 +69,59 @@ Mgr::QueryParams::find(const String& name) const
-        param.first = paramStr.substr(pmatch[1].rm_so, pmatch[1].rm_eo);
-        std::vector<int> array;
-        int n = pmatch[2].rm_so;
-        for (int i = n; i < pmatch[2].rm_eo; ++i) {
-            if (paramStr[i] == ',') {
-                array.push_back(atoi(paramStr.substr(n, i).termedBuf()));
-                n = i + 1;
-            }
-        }
-        if (n < pmatch[2].rm_eo)
-            array.push_back(atoi(paramStr.substr(n, pmatch[2].rm_eo).termedBuf()));
-        param.second = new IntParam(array);
-        parsed = true;
-    } else if (regexec(&stringExpr, paramStr.termedBuf(), 3, pmatch, 0) == 0) {
-        param.first = paramStr.substr(pmatch[1].rm_so, pmatch[1].rm_eo);
-        param.second = new StringParam(paramStr.substr(pmatch[2].rm_so, pmatch[2].rm_eo));
-        parsed = true;
+    static const CharacterSet nameChars = CharacterSet("param-name", "_") + CharacterSet::ALPHA + CharacterSet::DIGIT;
+
+    Parser::Tokenizer tok(paramStr);
+
+    SBuf name;
+    if (!tok.prefix(name, nameChars))
+        return false;
+
+    if (!tok.skip('='))
+        return false; // TODO: support name-only parameters
+
+    static const CharacterSet badValueChars = CharacterSet("param-value-bad", "&= ");
+    SBuf valueStr = tok.remaining();
+    if (valueStr.findFirstOf(badValueChars) != SBuf::npos)
+        return false;
+
+    param.first = SBufToString(name);
+
+    std::vector<int> array;
+    int64_t intVal = 0;
+    while (tok.int64(intVal, 10, false)) {
+        static const CharacterSet comma("comma",",");
+        (void)tok.skipAll(comma);
+        Must(intVal >= std::numeric_limits<int>::min());
+        Must(intVal <= std::numeric_limits<int>::max());
+        array.emplace_back(int(intVal));
     }
-    regfree(&stringExpr);
-    regfree(&intExpr);
-    return parsed;
+
+    if (tok.atEnd())
+        param.second = new IntParam(array);
+    else
+        param.second = new StringParam(SBufToString(valueStr));
+
+    return true;
 }
 
 bool
-Mgr::QueryParams::Parse(const String& aParamsStr, QueryParams& aParams)
+Mgr::QueryParams::Parse(Parser::Tokenizer &tok, QueryParams &aParams)
 {
-    if (aParamsStr.size() != 0) {
+    static const CharacterSet paramDelim("query-param","&");
+
+    SBuf foundStr;
+    while (tok.token(foundStr, paramDelim)) {
         Param param;
-        size_t n = 0;
-        size_t len = aParamsStr.size();
-        for (size_t i = n; i < len; ++i) {
-            if (aParamsStr[i] == '&') {
-                if (!ParseParam(aParamsStr.substr(n, i), param))
-                    return false;
-                aParams.params.push_back(param);
-                n = i + 1;
-            }
-        }
-        if (n < len) {
-            if (!ParseParam(aParamsStr.substr(n, len), param))
-                return false;
-            aParams.params.push_back(param);
-        }
+        if (!ParseParam(foundStr, param))
+            return false;
+        aParams.params.push_back(param);
     }
```

Comment:
Please try to refactor to reuse tokenizer:

```C++
while (!tok.atEnd())
    aParams.params.push_back(ParseParam(tok));
```

It may be easier to refactor this after fixing other problems.

Reply:
There is a subtle behaviour of tok.token() that is being relied on here to prune away sequences of suffix "&" delimiters without attempting to parse an empty parameter string between each one.

Reply:
AFAICT, Tokenizer::token() does not work the way you think it does (I can explain why but I do not want to blow up this discussion).

```C++
    /*
     * \return true if found a non-empty token followed by a delimiter
     */
    bool token(SBuf &returnedToken, const CharacterSet &delimiters);
```

AFAICT, the current code does not handle the last parameters correctly -- it leaves it behind and then returns false. Glad my refactoring request led to that bug discovery. We can return to this refactoring after the bug is fixed; the final code may look a bit different from my original snippet.

BTW, I think it is up to us whether to ignore/skip or reject empty parameters. Safely ignoring them is probably overall better (for URL-generating scripts), but we do not have to be that flexible if being flexible is difficult. I suspect it is not difficult.

Please test before re-requesting review.

Reply:
Ooh, good catch. Fixed by acdf250d2